### ISSUE TITLE
Investigate building docs from scratch

### ISSUE THREAD

**Issue by willmcgugan** [Textualize/textual#2629]
Investigate the experience of getting the docs to build from a fresh environment.

We've had a few issues that suggest there may be some problems there, related to dependancies.

**Comment by davep** [comment]
[Related discord thread](https://discord.com/channels/1026214085173461072/1033754296224841768/1110271538323783701).

Pretty sure this wasn't an issue as of #1726 making it into the repo; so it looks like something in the toolchain may have changed between then and now.

**Comment by paul-ollis** [comment]
Hi.
I am the person who identified this yesterday.
I have tried this in a venv with the same results.
After poking around, my understanding is that there are 2 missing mkdocs plugins - privacy and blog. My understanding is that they are both built-in plugins, but only (currently) available to mkdocs sponsors.

**Comment by davep** [comment]
> After poking around, my understanding is that there are 2 missing mkdocs plugins - privacy and blog.

Indeed, see #1726 -- this is what that work was for. It moved the dependency on the blog plugin, which is only available to mkdocs sponsors, into a separate branch of docs building `Makefile` targets (the blog plugin isn't useful to people generally, it's just for building out own docs which includes our blog). Without diving in, what seems to have happened here is that the `privacy` plugin's status seems to have changed between then and now (that's to say: I don't know that that's the case, but that's the first thing I'd check; back when doing that PR I tested building from scratch without access the the insiders' edition and had no problems).

**Comment by paul-ollis** [comment]
I tried deleting 'privacy' from mkdocs-offline.yml. That seems to allow 'make docs-build-offline' to work OK.

**Comment by davep** [comment]
As I recall (I'd need to go back over what I did in February), the problem with that is that the "offline" serving and building of docs was intended to be just that: possible to build and view without an Internet connection; and I seem to remember that the privacy plugin was important to enabling that.

I would expect removing it would remove the error, but it's likely that it reintroduces the issue that #1726 was mainly seeking to address (in part allowing GNU/Linux distributions to build and make available the docs without needing a working net connection).

Whatever the correct fix is here, that's an aspect we'll need to keep in mind.

**Comment by paul-ollis** [comment]
FYI. The offline docs I was able to generate appear to be pretty much fully functional, including the search feature. I seem to be able to use then OK without any network connection. Definitely good enough for my current needs.

The icon at the top left of each page:

![ksnip_20230523-135100](https://github.com/Textualize/textual/assets/6062510/c889ea9a-7eda-4f90-98ef-3f5122729e4a)

does not work - the link is a directory (I assume it is missing 'index.html'). Everything else I have tried so far works as I would expect.

**Comment by TomJGooding** [comment]
Probably a naive approach (and I'm not too confident with my regex!) but searching "https" in `docs` seems to confirm mainly just fonts from the excalidraw SVGs.

```shell
grep -r -P --exclude-dir=blog '(?<!(?:f"|\]\()|\[link=)https' docs | less
```

**Comment by squidfunk** [comment]
> I might be mistaken, but looking at the mkdocs-material [Insiders page](https://squidfunk.github.io/mkdocs-material/insiders/#goals-completed), it looks like the privacy plugin features won't be merged for general availability until the $20K funding goal? Perhaps that was changed since Feb, but couldn't see anything obvious from a quick browse of their repo.

@TomJGooding the built-in privacy plugin will be merged when the [$14k goal](https://squidfunk.github.io/mkdocs-material/insiders/#14000-goats-horn) is reached, which should cover all you need to bundle external assets ☺️ There are some additional features to the plugin that include optimization support (automatically optimize downloaded images) and handling of external links, that will be merged when the [$20k goal](https://squidfunk.github.io/mkdocs-material/insiders/#20000-jalapeno) is reached.

With [GitHub auto-cancelling all PayPal subscribers](https://twitter.com/squidfunk/status/1643539228574269443), we lost a significant amount of sponsors, but we hope to reach the $12k (blog plugin etc.) and $14k (privacy plugin etc.) goals within this year.

Feel free to mark this as OT.

**Comment by davep** [comment]
Looking into this, steps taken and what I've found so far:

1. Cloned a fresh copy of Textual `main`
2. Ran `make setup`
3. Ran `make docs-serve-offline`

The result is:

```
rm -rf .screenshot_cache
echo "INHERIT: mkdocs-offline.yml" > mkdocs-nav-offline.yml
grep -v "\- \"*[Bb]log" mkdocs-nav.yml >> mkdocs-nav-offline.yml
poetry run mkdocs serve --config-file mkdocs-nav-offline.yml
ERROR    -  Config value 'plugins': The "privacy" plugin is not installed
Aborted with 1 Configuration Errors!
make: *** [docs-serve-offline] Error 1
```

The current setup comes from #1726; the inclusion of `privacy` in the offline version of the documentation is done by 45af25a248afc509b697612283cf14844959f972 where I seem to have included it for the very specific purpose of helping with allowing the documentation to be built and used without a net connection; [this was one of the requests made that motivated these changes](https://github.com/Textualize/textual/issues/631#issuecomment-1413837026).

I was pretty certain that, after work on #1726 had mostly concluded, I'd tested this with a fresh environment where the mkdocs insiders code *wasn't* available; curiously we've also not had any complaints from downstream either. Perhaps I'm misremembering that I tested this, or somehow managed to accidentally test in an environment that wasn't as pristine as I'd thought.

Either way: while it looks like the privacy plugin is one of the ways to go to let third parties build and read documentation offline, and also do so in a way that doesn't require a net connection at all and so maximises user privacy, it obviously doesn't fulfil that role at the moment so I'll look at removing it and seeing what impact it has on the original requirement mentioned in #631.

Hopefully it doesn't turn out that "anyone can build" and "can be built 100% offline" are mutually exclusive desires.

**Comment by davep** [comment]
Sure enough, doing this:

```diff
diff --git a/mkdocs-offline.yml b/mkdocs-offline.yml
index 0d367af4..a86cf92e 100644
--- a/mkdocs-offline.yml
+++ b/mkdocs-offline.yml
@@ -2,7 +2,6 @@ INHERIT: mkdocs-common.yml
 
 plugins:
   offline:
-  privacy:
   exclude:
     glob:
       - "**/_template.md"
 ```

and then running `make docs-serve-offline` in a clean (non-mkdocs-insiders) environment means the docs can be loaded up and viewed (minus the blog, which is the really important thing here):

![Screenshot 2023-05-29 at 07 39 38](https://github.com/Textualize/textual/assets/28237/e1b419de-b053-4d1d-810c-3deefd4c4132)

At this point of course there are a good number of external resources being used, which means it's not exactly "offline". So this change solves this issue, for sure. The next thing to check, to at least know either way, is to conduct this exercise with no Internet connection and to see if the docs still build and can be viewed okay (in other words to be sure that the requests for online resources don't affect the docs in a way that makes them unusable).

**Comment by davep** [comment]
Attempting to run `make docs-serve-offline` without an Internet connection gets so far and then results in this:

```
URLError: <urlopen error [Errno 8] nodename nor servname provided, or not known>
```

From the output it's not clear what is emitting this. The final step before this was:

```
screenshotting 'docs/examples/widgets/tree.py'
```

**Comment by davep** [comment]
Turning on verbose logging for mkdocs, it looks like this might actually be down to mkdocstrings:

```
DEBUG    -  Running 1 `env` events
DEBUG    -  mkdocstrings: Creating inventory file objects.inv
DEBUG    -  mkdocstrings: Waiting for 2 inventory download(s)
URLError: <urlopen error [Errno 8] nodename nor servname provided, or not known>
```

I strongly suspect this is down to #1915 -- that shouldn't be too much of a problem, we could have it so that intersphinxing only happens if the online docs are being built, I guess (unless there's some way of telling mkdocstrings to fail gracefully here?).

**Comment by davep** [comment]
With d4eb0a0c20f125186ab3f286d5bd6535078b90e3 I've managed to run `make docs-serve-offline` to completion and browse the results. The opening page looks a wee bit different, but is still usable:

<img width="1337" alt="Screenshot 2023-05-29 at 08 35 45" src="https://github.com/Textualize/textual/assets/28237/74cc1610-5412-4c1e-b5ea-029dda30a217">

and of course "handwritten" fonts for the diagrams don't work as they're online:

<img width="1337" alt="Screenshot 2023-05-29 at 08 36 51" src="https://github.com/Textualize/textual/assets/28237/e64e4780-5b4d-4eba-9568-85851377b877">

<img width="1337" alt="Screenshot 2023-05-29 at 08 36 27" src="https://github.com/Textualize/textual/assets/28237/d943054b-6012-4e4f-a8dc-9946d098986b">

but overall I think this will do the job. One more test is to build for reading in the filesystem, rather than building to serve locally, while offline, but I suspect that's going to be fine also.

Overall I think this approach should solve this issue and also #2639. Anyone downstream should be able to build 100% offline, and also read 100% offline, albeit with an (expected and acceptable) degraded experience. Building offline but viewing online will mean that external resources will be pulled in; which seems fine under the circumstances.

Going to do a couple more rounds of tests and then PR.

**Comment by davep** [comment]
`make docs-build-offline` and then browsing in the filesystem working fine with no Internet connection.
